xclim/core/utils.py

```diff
@@ -208,37 +208,153 @@ def uses_dask(da):
+    """
+    arr_copy = arr.copy()
+    quantiles = np.array([per / 100.0 for per in percentiles])
+    return _nan_quantile(arr_copy, quantiles, axis, alpha, beta)
 
-    Returns
-    -------
-    np.array
-    """
-    if p is None:
-        p = [50]
-
-    nan_count = np.isnan(arr).sum(axis=-1)
-    out = np.moveaxis(np.percentile(arr, p, axis=-1), 0, -1)
-    nans = (nan_count > 0) & (nan_count < arr.shape[-1])
-    if np.any(nans):
-        out_mask = np.stack([nans] * len(p), axis=-1)
-        # arr1 = arr.reshape(int(arr.size / arr.shape[-1]), arr.shape[-1])
-        # only use nanpercentile where we need it (slow performance compared to standard) :
-        out[out_mask] = np.moveaxis(
-            np.nanpercentile(arr[nans], p, axis=-1), 0, -1
-        ).ravel()
-    return out
+
+def virtual_index_formula(
+    array_size: Union[int, np.array], quantiles: np.array, alpha: float, beta: float
+) -> np.array:
+    """
+    Compute the floating point indexes of an array for the linear interpolation of quantiles.
+
+    Note:
+        Compared to R, -1 is added because R array indexes start at 1 (0 for python)
+    """
+    return array_size * quantiles + (alpha + quantiles * (1 - alpha - beta)) - 1
+
+
+def gamma_formula(
+    val: Union[float, np.array], val_floor: Union[int, np.array]
+) -> Union[float, np.array]:
+    """
+    Compute the gamma (a.k.a 'm' or weight) for the linear interpolation of quantiles.
+    """
+    return val - val_floor
+
+
+def linear_interpolation_formula(
+    left: Union[float, np.array],
+    right: Union[float, np.array],
+    gamma: Union[float, np.array],
+) -> Union[float, np.array]:
+    """
+    Compute the linear interpolation weighted by gamma on each point of two same shape array.
+    """
+    return gamma * right + (1 - gamma) * left
+
+
+def _nan_quantile(
+    arr: np.array,
+    quantiles: np.array,
+    axis: int = 0,
+    alpha: float = 1.0,
+    beta: float = 1.0,
+) -> Union[float, np.array]:
+    """
+    Get the quantiles of the array for the given axis.
+    A linear interpolation is performed using alpha and beta.
+
+    By default alpha == beta == 1 which performs the 7th method of Hyndman&Fan.
+    with alpha == beta == 1/3 we get the 8th method.
+    """
+    # --- Setup
+    values_count = arr.shape[axis]
+    if values_count == 0:
+        return np.NAN
+    if values_count == 1:
+        result = np.take(arr, 0, axis=axis)
+        return np.broadcast_to(result, (quantiles.size,) + result.shape)
+    # The dimensions of `q` are prepended to the output shape, so we need the
+    # axis being sampled from `ap` to be first.
+    arr = np.moveaxis(arr, axis, 0)
+    # nan_count is not a scalar
+    nan_count = np.isnan(arr).sum(0).astype(float)
+    valid_values_count = values_count - nan_count
+    # We need at least two values to do an interpolation
+    too_few_values = valid_values_count < 2
+    if too_few_values.any():
```

Comment:
Don't you need to call `any` on a specific axis ?

Reply:
valid_values_count is derivative from the  valid_values_count which himself comes from a scalar (value_count) and a reduced array (nan_count) so I don't think we need an axis parameter.

However it might be a code smell that the execution flow is hard to follow and it needs some refactoring.